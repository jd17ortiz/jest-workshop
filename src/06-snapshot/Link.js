import React from 'react';

const STATUS = {
  HOVERED: 'hovered',
  NORMAL: 'normal',
};

export default class Link extends React.Component {
  constructor() {
    super();

    this._onMouseEnter = this._onMouseEnter.bind(this);
    this._onMouseLeave = this._onMouseLeave.bind(this);
    this._onClick = this._onClick.bind(this);

    this.state = {
      class: STATUS.NORMAL,
      msg: null
    };
  }

  _onMouseEnter() {
    this.setState({class: STATUS.HOVERED});
  }

  _onMouseLeave() {
    this.setState({class: STATUS.NORMAL});
  }

  _onClick(){
    this.setState({
      msg: 'already click'
    });
  }

  render() {
    let msg = this.state.msg? (<span>{this.state.msg}</span>): null;
    return (
      <div>
        <a
          className={this.state.class}
          href={this.props.page || '#'}
          onMouseEnter={this._onMouseEnter}
          onMouseLeave={this._onMouseLeave}
          onClick={this._onClick}
        >
        {this.props.children}
        </a>
        {msg}
      </div>
  );
  }
}