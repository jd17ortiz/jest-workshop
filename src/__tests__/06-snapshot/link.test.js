/* eslint-disable no-unused-vars */

"use strict";

import React from "react";
import Link from "../../06-snapshot/Link";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer
    .create(<Link page="http://www.facebook.com">Facebook</Link>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("changes the class when hovered", () => {
  const component = renderer.create(
    <Link page="http://www.google.com">Facebook</Link>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  tree.children[0].props.onMouseEnter();
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  tree.children[0].props.onMouseLeave();
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("msg is show when click link", () => {
  const component = renderer.create(
    <Link page="http://www.facebook.com">Facebook</Link>
  );
  let tree = component.toJSON();
  // manually trigger the callback
  tree.children[0].props.onClick();
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
