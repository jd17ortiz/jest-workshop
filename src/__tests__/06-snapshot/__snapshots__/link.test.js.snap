// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the class when hovered 1`] = `
<div>
  <a
    className="normal"
    href="http://www.facebook.com"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Facebook
  </a>
</div>
`;

exports[`changes the class when hovered 2`] = `
<div>
  <a
    className="hovered"
    href="http://www.facebook.com"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Facebook
  </a>
</div>
`;

exports[`changes the class when hovered 3`] = `
<div>
  <a
    className="normal"
    href="http://www.facebook.com"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Facebook
  </a>
</div>
`;

exports[`msg is show when click link 1`] = `
<div>
  <a
    className="normal"
    href="http://www.facebook.com"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Facebook
  </a>
  <span>
    already click
  </span>
</div>
`;

exports[`renders correctly 1`] = `
<div>
  <a
    className="normal"
    href="http://www.facebook.com"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Facebook
  </a>
</div>
`;
